Add math metric alarm

apechimp opened the pull request:
<!--- Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests --->
Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/6551

Changes proposed in this pull request:

* added `metric_query` to support adding [`MetricDataQuery`](https://docs.aws.amazon.com/sdk-for-go/api/service/cloudwatch/#MetricDataQuery)s to alarms via the `Metrics` property. An example follows.
```
resource "aws_cloudwatch_metric_alarm" "foobar" {
  alarm_name                = "terraform-test-foobar%d"
  comparison_operator       = "GreaterThanOrEqualToThreshold"
  evaluation_periods        = "2"
	threshold                 = "80"
  alarm_description         = "This metric monitors ec2 cpu utilization"
  insufficient_data_actions = []
	metric_query {
		id = "e1"
		expression = "m1"
		label = "cat"
		return_data = "true"
	}
	metric_query {
		id = "m1"
		metric {
			metric_name = "CPUUtilization"
			namespace   = "AWS/EC2"
			period      = "120"
			stat        = "Average"
			unit        = "Count"
			dimensions {
				InstanceId = "i-abc123"
			}
		}
	}
}
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCloudWatchMetricAlarm_expression'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSCloudWatchMetricAlarm_expression -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSCloudWatchMetricAlarm_expression
=== PAUSE TestAccAWSCloudWatchMetricAlarm_expression
=== CONT  TestAccAWSCloudWatchMetricAlarm_expression
--- PASS: TestAccAWSCloudWatchMetricAlarm_expression (20.02s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       20.063s

```

This is my first attempt at a PR to this repo, so apologies in advance for all the conventions I've violated 😉

matthiassplieth6666 commented:
@bflad any chance to get this merged?

humayunjamal commented:
when this can be merged ?

orthimnas commented:
I would also like to know when this can be merged.

amiteswar commented:
Seems like a lot of interest for this PR, When this can be merged? Thanks in advance.

arun251 commented:
This would be great to have.

shaiguitar commented:
Thanks @apechimp! Am looking forward to using this.

bflad commented:
Reviewing this on my to-do list this week unless another maintainer beats me to it.

bflad commented on aws/resource_aws_cloudwatch_metric_alarm.go:
```diff
@@ -187,6 +253,47 @@ func resourceAwsCloudWatchMetricAlarmRead(d *schema.ResourceData, meta interface
+			metricQueries[i] = metricQuery
+		}
+		if err := d.Set("metric_query", metricQueries); err != nil {
+			return err
```
We should provide error context for operators and code maintainers here, e.g.

```suggestion
			return fmt.Errorf("error setting metric_query: %s", err)
```

bflad commented on aws/resource_aws_cloudwatch_metric_alarm_test.go:
```diff
@@ -236,6 +236,26 @@ func TestAccAWSCloudWatchMetricAlarm_extendedStatistic(t *testing.T) {
+					testAccCheckCloudWatchMetricAlarmExists("aws_cloudwatch_metric_alarm.foobar", &alarm),
+					resource.TestCheckResourceAttr("aws_cloudwatch_metric_alarm.foobar", "metric_query.#", "2"),
+				),
+			},
```
Since the usage of expressions does not require recreating the Terraform resource, we should include another `TestStep` similar to the first which provides an updated test configuration (e.g. adding another `metric_query` configuration and including it in the expression), e.g.

```go
				Config: testAccAWSCloudWatchMetricAlarmConfigWithExpressionUpdated(rInt),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckCloudWatchMetricAlarmExists("aws_cloudwatch_metric_alarm.foobar", &alarm),
					resource.TestCheckResourceAttr("aws_cloudwatch_metric_alarm.foobar", "metric_query.#", "3"),
				),
```

As well as a `TestStep` that verifies these new configurations can be properly imported as the `aws_cloudwatch_metric_alarm` resource currently supports resource import, e.g.

```go
{
  ResourceName: "aws_cloudwatch_metric_alarm",
  ImportState: true,
  ImportStateVerify: true,
},
```

apechimp commented:
Thanks for the review, @bflad. I've addressed all the comments on the code and am currently working on the docs. Let me know if any more code changes are necessary.

bflad commented:
The code and testing looks great! This should be ready to merge once the docs are in and reviewed.

```
--- PASS: TestAccAWSCloudWatchMetricAlarm_missingStatistic (4.10s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_extendedStatistic (9.44s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm (9.52s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_basic (10.42s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction (11.40s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic (12.48s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_treatMissingData (14.27s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles (14.30s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_expression (16.20s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate (200.75s)
```

bflad commented on website/docs/r/cloudwatch_metric_alarm.html.markdown:
````diff
@@ -57,6 +57,51 @@ resource "aws_cloudwatch_metric_alarm" "bat" {
 }
 ```
 
+## Example with an Expression
````
😍

bflad commented:
This has been released in [version 1.59.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1590-february-14-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
